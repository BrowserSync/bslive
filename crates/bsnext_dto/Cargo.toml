[package]
name = "bsnext_dto"
version = "0.2.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
bsnext_input = { path = "../bsnext_input" }
bsnext_fs = { path = "../bsnext_fs" }
bsnext_tracing = { path = "../bsnext_tracing" }

serde = { workspace = true }
thiserror = { workspace = true }
serde_json = { workspace = true }
actix = { workspace = true }
typeshare = { workspace = true }
