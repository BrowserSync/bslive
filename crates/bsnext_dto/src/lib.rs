use actix::MessageResponse;
use bsnext_input::server_config::ServerIdentity;
use bsnext_input::InputError;
use std::fmt::{Display, Formatter};
use std::net::SocketAddr;
use std::path::Path;

use crate::internal::{ServerError, StartupEvent};
use bsnext_fs::Debounce;
use bsnext_input::client_config::ClientConfig;
use bsnext_input::route::{DirRoute, ProxyRoute, RawRoute, Route, RouteKind};
use bsnext_tracing::LogLevel;
use typeshare::typeshare;

pub mod external_events;
pub mod internal;
pub mod internal_events;
pub mod startup_events;

#[typeshare]
#[derive(Debug, serde::Serialize)]
pub struct ServerDesc {
    pub routes: Vec<RouteDTO>,
    pub id: String,
}

#[typeshare]
#[derive(Debug, serde::Serialize)]
pub struct RouteDTO {
    pub path: String,
    pub kind: RouteKindDTO,
}

impl From<Route> for RouteDTO {
    fn from(value: Route) -> Self {
        Self {
            path: value.path.as_str().to_owned(),
            kind: RouteKindDTO::from(value.kind),
        }
    }
}
impl From<&Route> for RouteDTO {
    fn from(value: &Route) -> Self {
        Self {
            path: value.path.as_str().to_owned(),
            kind: RouteKindDTO::from(value.kind.clone()),
        }
    }
}

#[typeshare]
#[derive(Debug, serde::Serialize)]
#[serde(tag = "kind", content = "payload")]
pub enum RouteKindDTO {
    Html { html: String },
    Json { json_str: String },
    Raw { raw: String },
    Sse { sse: String },
    Proxy { proxy: String },
    Dir { dir: String, base: Option<String> },
}

impl From<RouteKind> for RouteKindDTO {
    fn from(value: RouteKind) -> Self {
        match value {
            RouteKind::Raw(raw) => match raw {
                RawRoute::Html { html } => RouteKindDTO::Html { html },
                RawRoute::Json { json } => RouteKindDTO::Json {
                    json_str: serde_json::to_string(&json).expect("unreachable"),
                },
                RawRoute::Raw { raw } => RouteKindDTO::Raw { raw },
                RawRoute::Sse { sse } => RouteKindDTO::Sse { sse },
            },
            RouteKind::Proxy(ProxyRoute { proxy }) => RouteKindDTO::Proxy { proxy },
            RouteKind::Dir(DirRoute { dir, base }) => RouteKindDTO::Dir {
                dir,
                base: base.map(|b| b.to_string_lossy().to_string()),
            },
        }
    }
}

#[typeshare]
#[derive(Debug, Clone, serde::Serialize)]
pub struct ServersChangedDTO {
    pub servers_resp: GetActiveServersResponseDTO,
}

#[typeshare]
#[derive(Debug, serde::Serialize)]
pub enum EventLevel {
    #[serde(rename = "BSLIVE_EXTERNAL")]
    External,
}

#[typeshare]
#[derive(Debug, serde::Serialize)]
#[serde(tag = "kind", content = "payload")]
pub enum StartupEventDTO {
    Started,
    FailedStartup(String),
}

pub type StartupResult = Result<DidStart, StartupError>;

impl From<&StartupEvent> for StartupEventDTO {
    fn from(value: &StartupEvent) -> Self {
        match value {
            StartupEvent::Started => StartupEventDTO::Started,
            StartupEvent::FailedStartup(StartupError::InputError(err)) => {
                StartupEventDTO::FailedStartup(err.to_string())
            }
            StartupEvent::FailedStartup(StartupError::Other(err)) => {
                StartupEventDTO::FailedStartup(err.to_string())
            }
            StartupEvent::FailedStartup(StartupError::ServerError(err)) => {
                StartupEventDTO::FailedStartup(err.to_string())
            }
        }
    }
}

#[typeshare]
#[derive(Debug, serde::Serialize, Clone)]
pub struct InputAcceptedDTO {
    pub path: String,
}

#[typeshare]
#[derive(Debug, serde::Serialize, Clone)]
pub struct FileChangedDTO {
    pub path: String,
}

#[typeshare]
#[derive(Debug, serde::Serialize, Clone)]
pub struct FilesChangedDTO {
    pub paths: Vec<String>,
}

#[typeshare]
#[derive(Debug, Clone, serde::Serialize)]
pub struct WatchingDTO {
    pub paths: Vec<String>,
    pub debounce: DebounceDTO,
}

#[typeshare]
#[derive(Debug, Clone, serde::Serialize)]
pub struct StoppedWatchingDTO {
    pub paths: Vec<String>,
}

impl StoppedWatchingDTO {
    pub fn from_path_buf(p: &Path) -> Self {
        Self {
            paths: vec![p.to_string_lossy().to_string()],
        }
    }
}

impl Display for WatchingDTO {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "yo")
    }
}

#[typeshare]
#[derive(Debug, Clone, serde::Serialize)]
pub struct DebounceDTO {
    kind: String,
    ms: String,
}

impl Display for DebounceDTO {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}ms", self.kind, self.ms)
    }
}

impl From<Debounce> for DebounceDTO {
    fn from(value: Debounce) -> Self {
        match value {
            Debounce::Trailing { duration } => Self {
                kind: "trailing".to_string(),
                ms: duration.as_millis().to_string(),
            },
            Debounce::Buffered { duration } => Self {
                kind: "buffered".to_string(),
                ms: duration.as_millis().to_string(),
            },
        }
    }
}

impl FileChangedDTO {
    pub fn from_path_buf(p: &Path) -> Self {
        Self {
            path: p.to_string_lossy().to_string(),
        }
    }
}

impl WatchingDTO {
    pub fn from_path_buf(p: &Path, debounce: Debounce) -> Self {
        Self {
            paths: vec![p.to_string_lossy().to_string()],
            debounce: debounce.into(),
        }
    }
}

#[typeshare]
#[derive(Debug, Clone, serde::Serialize)]
#[serde(tag = "kind", content = "payload")]
pub enum ServerChange {
    Stopped { bind_address: String },
    Started,
    Patched,
    Errored { error: String },
}

#[typeshare]
#[derive(Debug, serde::Serialize)]
pub struct ServerChangeSetItem {
    pub identity: ServerIdentityDTO,
    pub change: ServerChange,
}
#[typeshare]
#[derive(Debug, serde::Serialize)]
pub struct ServerChangeSet {
    pub items: Vec<ServerChangeSetItem>,
}

#[derive(Debug, Clone, Default, MessageResponse)]
pub struct GetActiveServersResponse {
    pub servers: Vec<ActiveServer>,
}

// impl Default for GetActiveServersResponse {
//     fn default() -> Self {
//         Self { servers: vec![] }
//     }
// }

#[typeshare::typeshare]
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize, MessageResponse)]
pub struct GetActiveServersResponseDTO {
    pub servers: Vec<ServerDTO>,
}

impl From<&GetActiveServersResponse> for GetActiveServersResponseDTO {
    fn from(value: &GetActiveServersResponse) -> Self {
        Self {
            servers: value
                .servers
                .iter()
                .map(|s| ServerDTO {
                    id: s.identity.as_id().to_string(),
                    identity: (&s.identity).into(),
                    socket_addr: s.socket_addr.to_string(),
                })
                .collect(),
        }
    }
}

#[typeshare::typeshare]
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ServerDTO {
    pub id: String,
    pub identity: ServerIdentityDTO,
    pub socket_addr: String,
}

impl From<&ActiveServer> for ServerDTO {
    fn from(value: &ActiveServer) -> Self {
        Self {
            id: value.identity.as_id().to_string(),
            identity: ServerIdentityDTO::from(&value.identity),
            socket_addr: value.socket_addr.to_string(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct ActiveServer {
    pub identity: ServerIdentity,
    pub socket_addr: SocketAddr,
}

#[derive(Debug)]
pub enum DidStart {
    Started(GetActiveServersResponse),
}

#[derive(Debug, thiserror::Error)]
pub enum StartupError {
    #[error("{0}")]
    InputError(#[from] InputError),
    #[error("{0}")]
    ServerError(#[from] ServerError),
    #[error("{0}")]
    Other(String),
}

#[typeshare::typeshare]
#[derive(Debug, PartialEq, Hash, Eq, Clone, serde::Deserialize, serde::Serialize)]
#[serde(tag = "kind", content = "payload")]
pub enum ServerIdentityDTO {
    Both { name: String, bind_address: String },
    Address { bind_address: String },
    Named { name: String },
}

impl From<&ServerIdentity> for ServerIdentityDTO {
    fn from(value: &ServerIdentity) -> Self {
        match value {
            ServerIdentity::Both { name, bind_address } => ServerIdentityDTO::Both {
                name: name.to_owned(),
                bind_address: bind_address.to_owned(),
            },
            ServerIdentity::Address { bind_address } => ServerIdentityDTO::Address {
                bind_address: bind_address.to_owned(),
            },
            ServerIdentity::Named { name } => ServerIdentityDTO::Named {
                name: name.to_owned(),
            },
        }
    }
}

#[typeshare::typeshare]
#[derive(Debug, PartialEq, Hash, Eq, Clone, serde::Deserialize, serde::Serialize)]
#[serde(tag = "kind", content = "payload")]
pub enum InputErrorDTO {
    MissingInputs(String),
    InvalidInput(String),
    NotFound(String),
    InputWriteError(String),
    PathError(String),
    PortError(String),
    DirError(String),
    YamlError(String),
    MarkdownError(String),
    HtmlError(String),
    Io(String),
    UnsupportedExtension(String),
    MissingExtension(String),
    EmptyInput(String),
    BsLiveRules(String),
}

impl From<&InputError> for InputErrorDTO {
    fn from(value: &InputError) -> Self {
        match value {
            e @ InputError::MissingInputs => InputErrorDTO::MissingInputs(e.to_string()),
            e @ InputError::InvalidInput(_) => InputErrorDTO::InvalidInput(e.to_string()),
            e @ InputError::NotFound(_) => InputErrorDTO::NotFound(e.to_string()),
            e @ InputError::InputWriteError(_) => InputErrorDTO::InputWriteError(e.to_string()),
            e @ InputError::PathError(_) => InputErrorDTO::PathError(e.to_string()),
            e @ InputError::PortError(_) => InputErrorDTO::PortError(e.to_string()),
            e @ InputError::DirError(_) => InputErrorDTO::DirError(e.to_string()),
            e @ InputError::MarkdownError(_) => InputErrorDTO::MarkdownError(e.to_string()),
            e @ InputError::HtmlError(_) => InputErrorDTO::HtmlError(e.to_string()),
            e @ InputError::YamlError(_) => InputErrorDTO::YamlError(e.to_string()),
            e @ InputError::Io(_) => InputErrorDTO::Io(e.to_string()),
            e @ InputError::UnsupportedExtension(_) => {
                InputErrorDTO::UnsupportedExtension(e.to_string())
            }
            e @ InputError::MissingExtension(_) => InputErrorDTO::MissingExtension(e.to_string()),
            e @ InputError::EmptyInput => InputErrorDTO::EmptyInput(e.to_string()),
            e @ InputError::BsLiveRules(..) => InputErrorDTO::BsLiveRules(e.to_string()),
        }
    }
}

impl From<InputError> for InputErrorDTO {
    fn from(value: InputError) -> Self {
        (&value).into()
    }
}

#[typeshare::typeshare]
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(tag = "kind", content = "payload")]
pub enum ClientEvent {
    Change(ChangeDTO),
    WsConnection(ClientConfigDTO),
    Config(ClientConfigDTO),
}

#[typeshare::typeshare]
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct ClientConfigDTO {
    pub log_level: LogLevelDTO,
}

impl From<ClientConfig> for ClientConfigDTO {
    fn from(value: ClientConfig) -> Self {
        Self {
            log_level: value.log.into(),
        }
    }
}

impl From<&ClientConfig> for ClientConfigDTO {
    fn from(value: &ClientConfig) -> Self {
        Self {
            log_level: value.log.into(),
        }
    }
}

#[typeshare::typeshare]
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum LogLevelDTO {
    Info,
    Debug,
    Trace,
    Error,
}

impl From<LogLevel> for LogLevelDTO {
    fn from(value: LogLevel) -> Self {
        match value {
            LogLevel::Info => LogLevelDTO::Info,
            LogLevel::Debug => LogLevelDTO::Debug,
            LogLevel::Trace => LogLevelDTO::Trace,
            LogLevel::Error => LogLevelDTO::Error,
        }
    }
}

#[typeshare::typeshare]
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(tag = "kind", content = "payload")]
pub enum ChangeDTO {
    Fs {
        path: String,
        change_kind: ChangeKind,
    },
    FsMany(Vec<ChangeDTO>),
}

#[typeshare::typeshare]
#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
pub enum ChangeKind {
    Changed,
    Added,
    Removed,
}

#[typeshare::typeshare]
#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
pub struct InjectConfig {
    pub connect: ConnectInfo,
    pub ctx_message: String,
}

#[typeshare::typeshare]
#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
pub struct ConnectInfo {
    pub ws_path: String,
    pub host: Option<String>,
}
