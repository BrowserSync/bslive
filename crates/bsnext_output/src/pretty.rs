use crate::OutputWriter;
use bsnext_dto::internal::{ChildResult, InternalEvents};
use bsnext_dto::{
    ExternalEvents, FileChanged, FilesChangedDTO, IdentityDTO, InputAccepted, InputErrorDTO,
    ServersChanged, StartupErrorDTO, StartupEvent, StoppedWatching, Watching,
};
use std::io::Write;
use std::marker::PhantomData;
use std::path::PathBuf;

pub struct PrettyPrint;

impl OutputWriter for PrettyPrint {
    fn handle_external_event<W: Write>(
        &self,
        sink: &mut W,
        evt: &ExternalEvents,
    ) -> anyhow::Result<()> {
        match &evt {
            ExternalEvents::ServersChanged(servers_started) => {
                print_servers_changed(sink, servers_started)
            }
            ExternalEvents::InputError(input_err) => {
                print_input_error(sink, Indent::None, input_err)
            }
            ExternalEvents::Watching(watching) => print_watching(sink, watching),
            ExternalEvents::WatchingStopped(watching) => print_stopped_watching(sink, watching),
            ExternalEvents::InputAccepted(input_accepted) => {
                print_input_accepted(sink, input_accepted)
            }
            ExternalEvents::FileChanged(file_changed) => print_file_changed(sink, file_changed),
            ExternalEvents::FilesChanged(files_changed) => print_files_changed(sink, files_changed),
            ExternalEvents::InputFileChanged(file_changed) => {
                print_input_file_changed(sink, file_changed)
            }
        }
    }

    fn handle_internal_event<W: Write>(
        &self,
        sink: &mut W,
        evt: InternalEvents,
    ) -> anyhow::Result<()> {
        match evt {
            InternalEvents::ServersChanged {
                server_resp,
                child_results: _,
            } => print_servers_changed(
                sink,
                &ServersChanged {
                    servers_resp: server_resp,
                },
            ),
        }
    }

    fn handle_startup_event<W: Write>(
        &self,
        sink: &mut W,
        evt: &StartupEvent,
    ) -> anyhow::Result<()> {
        match evt {
            StartupEvent::Started => {
                writeln!(sink, "{}", Line::prefixed().info("started..."))?;
            }
            StartupEvent::FailedStartup(err) => {
                writeln!(
                    sink,
                    "{}",
                    Line::prefixed().info("An error prevented startup!")
                )?;
                writeln!(sink)?;
                match err {
                    StartupErrorDTO::InputError(input_err) => {
                        print_input_error(sink, Indent::Some(4), input_err)?;
                    }
                }
            }
        }
        Ok(())
    }
}

// const prefix: ANSIGenericString<str> = ansi_term::Color::Red.paint("[bslive]");

trait LineState {}
struct Line<T: LineState = Orig> {
    indent: Indent,
    _state: PhantomData<T>,
}
struct Orig;
impl LineState for Orig {}
struct Prefixed;
impl LineState for Prefixed {}

struct Unprefixed;
impl LineState for Unprefixed {}
impl Line<Orig> {
    pub fn prefixed() -> Line<Prefixed> {
        Line {
            indent: Indent::None,
            _state: PhantomData,
        }
    }
}
impl Line<Orig> {
    pub fn unprefixed() -> Line<Unprefixed> {
        Line {
            indent: Indent::None,
            _state: PhantomData,
        }
    }
}
impl Line<Prefixed> {
    pub fn info(self, str: &str) -> String {
        format!("[bslive] {}", ansi_term::Color::Cyan.paint(str))
    }
}
impl Line<Unprefixed> {
    pub fn indent(self, size: Indent) -> Self {
        Self {
            indent: size,
            _state: PhantomData,
        }
    }
    pub fn error(self, str: &str) -> String {
        let coloured = ansi_term::Color::Red.paint(str);
        indent::indent_all_by(self.indent.indent_size(), coloured.to_string())
    }
}

pub fn print_file_changed<W: Write>(w: &mut W, evt: &FileChanged) -> anyhow::Result<()> {
    writeln!(w, "[change] {}", evt.path)?;
    Ok(())
}

pub fn print_files_changed<W: Write>(w: &mut W, evt: &FilesChangedDTO) -> anyhow::Result<()> {
    match evt.paths.len() {
        0..=2 => {
            writeln!(w, "[change:multi] {}", short_file_list(&evt.paths))?;
        }
        3.. => {
            let other = evt.paths.len() - 2;
            let subset = evt.paths.iter().take(2).collect::<Vec<_>>();
            writeln!(
                w,
                "[change:multi] {} (and {} others)",
                short_file_list(&subset),
                other
            )?;
        }
    }
    Ok(())
}

fn short_file_list<A: AsRef<str>>(paths: &[A]) -> String {
    let file_names = paths
        .iter()
        .filter_map(|p| {
            PathBuf::from(p.as_ref())
                .file_name()
                .map(|filename| filename.to_string_lossy().to_string())
        })
        .collect::<Vec<_>>();
    file_names.join(", ")
}

pub fn print_input_file_changed<W: Write>(w: &mut W, evt: &FileChanged) -> anyhow::Result<()> {
    writeln!(w, "[change:input] {}", evt.path)?;
    Ok(())
}

pub fn print_input_accepted<W: Write>(w: &mut W, evt: &InputAccepted) -> anyhow::Result<()> {
    writeln!(w, "[input] {}", evt.path)?;
    Ok(())
}

pub fn print_watching<W: Write>(w: &mut W, evt: &Watching) -> anyhow::Result<()> {
    for x in &evt.paths {
        writeln!(w, "[watching {}] {}", evt.debounce, x)?;
    }
    Ok(())
}

enum Indent {
    None,
    Some(usize),
}

impl Indent {
    pub fn indent_size(&self) -> usize {
        match self {
            Indent::None => 0,
            Indent::Some(size) => *size,
        }
    }
}

fn print_input_error<W: Write>(
    w: &mut W,
    indent: Indent,
    evt: &InputErrorDTO,
) -> anyhow::Result<()> {
    let v = match evt {
        InputErrorDTO::MissingInputs(evt) => evt,
        InputErrorDTO::InvalidInput(evt) => evt,
        InputErrorDTO::NotFound(evt) => evt,
        InputErrorDTO::InputWriteError(evt) => evt,
        InputErrorDTO::PathError(evt) => evt,
        InputErrorDTO::PortError(evt) => evt,
        InputErrorDTO::DirError(evt) => evt,
        InputErrorDTO::YamlError(evt) => evt,
        InputErrorDTO::MarkdownError(evt) => evt,
        InputErrorDTO::Io(evt) => evt,
        InputErrorDTO::UnsupportedExtension(evt) => evt,
        InputErrorDTO::MissingExtension(evt) => evt,
        InputErrorDTO::EmptyInput(evt) => evt,
    };
    writeln!(w, "{}", Line::unprefixed().indent(indent).error(v))?;
    Ok(())
}

pub fn print_stopped_watching<W: Write>(w: &mut W, evt: &StoppedWatching) -> anyhow::Result<()> {
    for x in &evt.paths {
        writeln!(w, "[watching:stopped] {}", x)?;
    }
    Ok(())
}

fn print_servers_changed<W>(w: &mut W, servers_started: &ServersChanged) -> anyhow::Result<()>
where
    W: Write,
{
    let ServersChanged {
        servers_resp,
        // changeset,
    } = servers_started;

    for server_dto in &servers_resp.servers {
        match &server_dto.identity {
            IdentityDTO::Both { name, .. } => {
                writeln!(w, "[server] [{}] http://{}", name, server_dto.socket_addr)?;
            }
            IdentityDTO::Address { .. } => {
                writeln!(w, "[server] http://{}", server_dto.socket_addr)?;
            }
            IdentityDTO::Named { name } => {
                writeln!(w, "[server] [{}] http://{}", name, &server_dto.socket_addr)?
            }
        }
    }

    // for ServerChangeSetItem { change, identity } in &changeset.items {
    //     let running = servers_resp
    //         .servers
    //         .iter()
    //         .find(|x| x.identity == *identity);
    //     match change {
    //         ServerChange::Stopped { bind_address } => match &identity {
    //             IdentityDTO::Both { name, bind_address } => {
    //                 writeln!(w, "[server removed] [{name}] http://{bind_address}")?;
    //             }
    //             IdentityDTO::Address { bind_address } => {
    //                 writeln!(w, "[server removed] http://{bind_address}")?;
    //             }
    //             IdentityDTO::Named { name } => {
    //                 writeln!(w, "[server removed] [{name}] http://{}", bind_address)?;
    //             }
    //         },
    //         ServerChange::Started => match &identity {
    //             IdentityDTO::Both { name, bind_address } => {
    //                 if running.is_some() {
    //                     writeln!(w, "[server added] [{}] http://{}", name, bind_address)?;
    //                 }
    //             }
    //             IdentityDTO::Address { bind_address } => {
    //                 if running.is_some() {
    //                     writeln!(w, "[server added] http://{}", bind_address)?;
    //                 }
    //             }
    //             IdentityDTO::Named { name } => {
    //                 if let Some(running) = running {
    //                     writeln!(
    //                         w,
    //                         "[server added] [{}] http://{}",
    //                         name, &running.socket_addr
    //                     )?;
    //                 } else {
    //                     unreachable!("?");
    //                 }
    //             }
    //         },
    //         ServerChange::Patched => {}
    //         ServerChange::Errored { error } => {
    //             writeln!(w, "[❌ server failed] {} {}", iden(identity), error)?;
    //         }
    //     }
    // }
    Ok(())
}

pub fn print_server_updates(evts: &[ChildResult]) -> Vec<String> {
    evts.iter()
        .map(|r| match r {
            ChildResult::Created(created) => {
                vec![format!(
                    "[created] {}",
                    server_display(
                        &IdentityDTO::from(&created.server_handler.identity),
                        &created.server_handler.socket_addr.to_string()
                    ),
                )]
            }
            ChildResult::Stopped(stopped) => {
                vec![format!("[stopped] {}", iden(stopped))]
            }
            ChildResult::CreateErr(errored) => {
                vec![format!(
                    "[server] errored... {:?} {} ",
                    iden(&errored.identity),
                    errored.server_error
                )]
            }
            ChildResult::Patched(child) => {
                let mut lines = vec![];
                // todo: determine WHICH changes were actually applied (instead of saying everything was patched)
                for x in &child.route_change_set.changed {
                    lines.push(format!(
                        "[patched] {} {:?}",
                        iden(&child.server_handler.identity),
                        x
                    ));
                }
                for x in &child.route_change_set.added {
                    lines.push(format!(
                        "[patched] {} {:?}",
                        iden(&child.server_handler.identity),
                        x
                    ));
                }
                lines
            }
            ChildResult::PatchErr(errored) => {
                vec![format!(
                    "[patch] error {} {} ",
                    iden(&errored.identity),
                    errored.patch_error
                )]
            }
        })
        .flatten()
        .collect()
}

pub fn iden(identity_dto: impl Into<IdentityDTO>) -> String {
    match identity_dto.into() {
        IdentityDTO::Both { name, bind_address } => format!("[{name}] {bind_address}"),
        IdentityDTO::Address { bind_address } => bind_address.to_string(),
        IdentityDTO::Named { name } => format!("[{name}]"),
    }
}

pub fn server_display(identity_dto: &IdentityDTO, socket_addr: &str) -> String {
    match &identity_dto {
        IdentityDTO::Both { name, .. } => {
            format!("[server] [{}] http://{}", name, socket_addr)
        }
        IdentityDTO::Address { .. } => {
            format!("[server] http://{}", socket_addr)
        }
        IdentityDTO::Named { name } => {
            format!("[server] [{}] http://{}", name, &socket_addr)
        }
    }
}
