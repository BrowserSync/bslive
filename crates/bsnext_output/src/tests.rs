use crate::pretty::PrettyPrint;
use crate::OutputWriter;
use bsnext_dto::{
    ExternalEvents, GetServersMessageResponse, IdentityDTO, ServerChange,
    ServerChangeSet, ServerChangeSetItem, ServerDTO, ServersStarted, StartupEvent,
};
use std::io::{BufWriter};

fn iden_1() -> IdentityDTO {
    IdentityDTO::Address {
        bind_address: "0.0.0.0:3000".to_string(),
    }
}
fn server_1() -> ServerDTO {
    ServerDTO {
        id: "abc".to_string(),
        identity: iden_1(),
        socket_addr: "0.0.0.0:3000".to_string(),
    }
}
fn iden_2() -> IdentityDTO {
    IdentityDTO::Address {
        bind_address: "0.0.0.0:4000".to_string(),
    }
}
fn server_2() -> ServerDTO {
    ServerDTO {
        id: "abcdef".to_string(),
        identity: iden_1(),
        socket_addr: "0.0.0.0:4000".to_string(),
    }
}

fn exec(evt: &ExternalEvents) -> anyhow::Result<String> {
    let mut writer = BufWriter::new(Vec::new());
    PrettyPrint.handle_external_event(&mut writer, &evt)?;
    Ok(String::from_utf8(writer.into_inner()?).unwrap())
}

fn exec_startup(evt: &StartupEvent) -> anyhow::Result<(String, String)> {
    let mut writer = BufWriter::new(Vec::new());
    PrettyPrint.handle_startup_event(&mut writer, &evt)?;
    let raw = String::from_utf8(writer.into_inner()?).unwrap();
    let stripped = strip_ansi_escapes::strip(&raw);
    let stripped_str = std::str::from_utf8(&stripped).unwrap().to_string();
    Ok((raw, stripped_str))
}

#[test]
fn test_servers_started() -> anyhow::Result<()> {
    let evt = ExternalEvents::ServersStarted(ServersStarted {
        servers_resp: GetServersMessageResponse {
            servers: vec![server_1(), server_2()],
        },
        changeset: ServerChangeSet {
            items: vec![
                ServerChangeSetItem {
                    identity: iden_1(),
                    change: ServerChange::Started,
                },
                ServerChangeSetItem {
                    identity: iden_2(),
                    change: ServerChange::Stopped {
                        bind_address: server_2().socket_addr,
                    },
                },
            ],
        },
    });
    let actual = exec(&evt).unwrap();
    insta::assert_snapshot!(actual);
    Ok(())
}

#[test]
fn test_startup_success() -> anyhow::Result<()> {
    let evt = StartupEvent::Started;
    let (actual, stripped) = exec_startup(&evt).unwrap();
    insta::assert_snapshot!(&actual);
    insta::assert_snapshot!(stripped);
    Ok(())
}
