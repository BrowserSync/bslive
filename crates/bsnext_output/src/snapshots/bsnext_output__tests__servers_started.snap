---
source: crates/bsnext_output/src/tests.rs
expression: actual
---
[server] http://0.0.0.0:3000
[server] http://0.0.0.0:4000
